compare/compare.c

```diff
@@ -0,0 +1,639 @@
+}
+
+
+static int print_usage(const char *program)
+{
+    const char *fmt =
+        "Usage: %s [-i] <input-dir> [-o <output-dir>] [-s <scale:1-8>]\n";
+    printf(fmt, program);
+    return 1;
+}
+
+void msg_callback(int level, const char *fmt, va_list va, void *data)
+{
+    if (level > 3)
+        return;
+    printf("libass: ");
+    vprintf(fmt, va);
+    printf("\n");
+}
+
+int main(int argc, char *argv[])
+{
+    enum {
+        INPUT, OUTPUT, SCALE
+    };
+    int pos[3] = {0};
+    for (int i = 1; i < argc; i++) {
+        if (argv[i][0] != '-') {
+            if (pos[INPUT])
+                return print_usage(argv[0]);
+            pos[INPUT] = i;
+            continue;
+        }
+        int index;
+        switch (argv[i][1]) {
+        case 'i':  index = INPUT;   break;
+        case 'o':  index = OUTPUT;  break;
+        case 's':  index = SCALE;   break;
+        default:  return print_usage(argv[0]);
+        }
+        if (argv[i][2] || ++i >= argc || pos[index])
+            return print_usage(argv[0]);
+        pos[index] = i;
+    }
+    if (!pos[INPUT])
+        return print_usage(argv[0]);
+
+    int scale = 1;
+    if (pos[SCALE]) {
+        const char *arg = argv[pos[SCALE]];
+        if (arg[0] < '1' || arg[0] > '8' || arg[1]) {
+            printf("Invalid scale value, should be 1-8!\n");
+            return 1;
+        }
+        scale = arg[0] - '0';
+    }
+
+    const char *input = argv[pos[INPUT]];
+    DIR *dir = opendir(input);
+    if (!dir) {
+        printf("Cannot open input directory '%s'!\n", input);
+        return 1;
+    }
+
+    const char *output = NULL;
+    if (pos[OUTPUT]) {
+        output = argv[pos[OUTPUT]];
+        struct stat st;
+        if (stat(output, &st)) {
+            if (mkdir(output, 0755)) {
+                printf("Cannot create output directory '%s'!\n", output);
+                closedir(dir);
+                return 1;
+            }
+        } else if (!(st.st_mode & S_IFDIR)) {
+            printf("Invalid output directory '%s'!\n", output);
+            closedir(dir);
+            return 1;
+        }
+    }
+
+    ASS_Library *lib = ass_library_init();
+    if (!lib) {
+        printf("ass_library_init failed!\n");
+        closedir(dir);
+        return 1;
+    }
+    ass_set_message_cb(lib, msg_callback, NULL);
+
+    ItemList list;
+    if (!init_items(&list)) {
+        ass_library_done(lib);
+        closedir(dir);
+        return 1;
+    }
+
+#define COMBINE(a, b, c, d) \
+    ((uint32_t) (uint8_t) (a) | \
+    ((uint32_t) (uint8_t) (b) | \
+    ((uint32_t) (uint8_t) (c) | \
+    ((uint32_t) (uint8_t) (d)) << 8) << 8) << 8)
+
+    while (true) {
+        struct dirent *file = readdir(dir);
+        if (!file)
+            break;
+        const char *name = file->d_name;
+        if (name[0] == '.')
+            continue;
+        int len = strlen(name);
+        if (len <= 4)
+            continue;
+        switch (COMBINE(name[len - 4], name[len - 3],
```

Comment:
This is overkill. Just use `strrchr` and `strcmp`; it's much clearer.

Reply:
+1

Reply:
I've seen many cases where chained string comparisons lead to significant performance degradation (and it's long-standing problem of libass parsing) so I consider it bad practice and avoid it when I can. Is that specific chunk of code hard to understand?

Reply:
That may be true for subtitle parsing, but it's complete overkill in this case, and it's confusing. It'll also be obnoxious if someone wants to match a different number of characters, or decides to make the matching case-insensitive. Regarding bad practice, I think code should primarily be easy to read, and there need to be good reasons to do something else.

Reply:
Yeah, the subtitle parsing code is a hot tight loop, and this isn't.